components/browser/session/src/main/java/mozilla/components/browser/session/engine/middleware/LastAccessMiddleware.kt

```diff
@@ -0,0 +1,35 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.session.engine.middleware
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.action.TabSessionAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+
+/**
+ * [Middleware] that handles updating the [TabSessionState.lastAccess] when a tab is selected.
+ */
+internal class LastAccessMiddleware : Middleware<BrowserState, BrowserAction> {
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        if (action is TabListAction.SelectTabAction) {
+            context.dispatchUpdateActionForId(action.tabId)
+        } else if (action is TabListAction.AddTabAction && action.select) {
+            context.dispatchUpdateActionForId(action.tab.id)
+        }
```

Comment:
nit: Just a nit... should we first select/add (invoking `next`) and *then* update the time? Especially the "add tab" case. How does this work if you update before the tab is added?

Reply:
I haven't noticed any side-effects in my testing, but I think your suggestion makes a lot more sense than the current behaviour.